pandas/io/json/json.py

```diff
@@ -686,7 +686,7 @@ def _try_convert_data(self, name, data, use_dtypes=True,
 
         result = False
 
-        if data.dtype == 'object':
+        if len(data) and data.dtype == 'object':
```

Comment:
This is the fix that seem to solve the error.

Any thoughts on this?

Reply:
The point of the JSON table schema is that we can be explicit about the types of the columns, so we shouldn't need to infer anything. Any way to avoid a call to this method altogether?

Reply:
> The point of the JSON table schema is that we can be explicit about the types of the columns, so we shouldn't need to infer anything.

@WillAyd : I'm confused how this is relevant to the patch.

Reply:
My point is that since the datatypes are explicitly defined in the schema that we shouldn't need any type of inference, which I get the impression this method is doing

Reply:
> My point is that since the datatypes are explicitly defined in the schema that we shouldn't need any type of inference, which I get the impression this method is doing

Sure, but I'm still not seeing relevance.  The patch is pretty straightforward.

Reply:
The relevance is this "fix" still wouldn't cover all use cases. It helps for the empty DataFrame, but what happens if the frame contains the value `1.` with a dtype of object? AFAICT would still cast to float, which again is not desirable and works against the explicitness of the JSON schema

Reply:
Yeah, to me it sounds like the possible problematics related to the method itself calls for a new issue, don’t you think?

Reply:
> Yeah, to me it sounds like the possible problematics related to the method itself calls for a new issue, don’t you think?

@pyryjook : Yes, I agree with this sentiment.

> The relevance is this "fix" still wouldn't cover all use cases. It helps for the empty DataFrame, but what happens if the frame contains the value 1. with a dtype of object?

@WillAyd : How is this concern relevant to the particular issue at hand, and wouldn't this have been a problem even before this patch.  We want PR's to be atomic, and I sense the ask here is a little too much for this PR.

Reply:
Agreed that it is conceptually two different issues tied together. I think there's three ways to go about this:

1. Implement this without the "len() fix", knowing that empty DataFrames that can be coerced to float will, just like non-empty DataFrames will
2. Implement this with the "len() fix" and live with the fact that empty DataFrames can be round-tripped cleanly but non-empty DataFrames with "float-like" data may still be undesirably coerced to float OR
3. Find a way to prevent the coercion altogether along with this change

I'm not clear what the level of effort is for the third point above. If it is large then for sure it would make sense for one of the first two options. Otherwise I don't think it makes sense to introduce this inconsistency into the code base, as nuanced as it is

Reply:
My personal opinion, I don't think option 2 makes much sense. As a user, I would rather have consistent, though undesirable behavior (and leave the test a bit weak, ignoring dtypes, or perhaps marking as expected failure?) than have empty and non empty data frame behave differently.
If at least non empty dataframe behaved as expected, and there was bad behavior on the corner case of empty ones, i could better see the case for living with inconsistent behavior.

Option 3 is obviously the best choice, though spontaneously seems a bit overkill to me? But, given #21140 I would be glad to take a look this week end and circle back with my best shot.

Reply:
I swiftly went thru the past commits related to the method and that way tried to find out the reasoning behind the implementation. First of all it's quite old method (5 years) and it has seen only a few modifications during it's lifetime. On top of that, the functionality of it does seem to be quite complexly tied to multiple use cases. 

On that light, knowing its quite subtle look on the matter, it sure looks like quite fundamental task to re-think the purpose (or existence) of that method within this PR.

I completely get the point that the fix with the `len()` or without would only be a compromise when looking at the whole picture.

Clearly, my opinion of the complexity is affected by the fact that this is my first contribution to this library :)

Reply:
I agree with @ludaavics that 2 is the least desirable. I would say if you don't see an apparent solution to number 3 above then create a separate issue about the coercion of object types to float with `read_json` and `table='orient'`. After that you can parametrize the test for `check_dtypes`, xfailing the strict check and placing a reference via a TODO comment to the issue around coercion.

Reply:
@WillAyd : Yes, that sounds like a good plan.

Reply:
Great, sounds like a plan! I'll make the new issue and the changes to the code accordingly. Thanks guys!